### DISCUSSION TITLE
Welcome to RAWSim-O Discussions!

### DISCUSSION THREAD

**Discussion by merschformann** [merschformann/RAWSim-O#4] [General]
## 👋 Welcome!
  Feel free to use [discussions](https://github.com/merschformann/RAWSim-O/discussions) for questions / ideas that do not immediately need to go into an issue. :hugs:

**Comment by ejpodnox** [comment]
Thanks for your nice work.  I wonder if you have the python interface for development?

**Reply by merschformann** [reply]
Hi @BillyCX ! Unfortunately, there is no such thing as a python interface. Could you elaborate a little what you have in mind? What should such an interface feature / what is the use case? Is there maybe a comparable simulation model out there that supports python bindings?
Probably, I won't have enough time to support something like this. 😞 However, if I ever find the time to do a rework of RAWSim-O (nowadays working on Linux only, therefore, necessary for myself), I would love to hear about such ideas (maybe I can address them).

Judging by your profile I would guess you are interested in doing some Machine Learning for RMFS or similar systems? Would it be of interest to have bindings for certain decision problems occurring in an RMFS? E.g., decide what task a robot should do next, decide in which pod/rack to store new inventory, etc.

Otherwise, I hope the code here still helps in one way or the other. If you are looking into such ML based decision making (just an assumption), you could also simply call your decider from RAWSim-O code. Of course, this is not very convenient. 😅 

Have a great day!

**Comment by LFK01** [comment]
Hi Marius,
I'm developing my master thesis @ Politecnico di Milano on Fault Detection and Diagnosis in Swarm Robotics. My professor suggested I look into your "Decision rules for robotic mobile fulfillment systems" paper. I immediately proceeded on looking at the RAWSim-O simulator and found it very interesting and well developed. You did an amazing job.

I'm not used to C# and Visual Studio at all but I managed to compile the project and tried some functionalities like in the tutorial provided in the README. I tried to implement a modified PodStorage controller like in the tutorial but I didn't succeed.

I'm here to ask for some help in developing some functionalities in the project and learning how to explore and understand your codebase. Is this the right space where to ask this?

To proceed with the development of my thesis, I need to inject faults in single robots and retrieve each bot position for each timestep. I tried to modify lines 711 and 712 of RAWSimO.Core.Bots.BotNormal.cs to:
```
XVelocity = 0;
YVelocity = 0;
```
and then recompiled the project. Compilation turned out successful but bots kept moving.

For now, I would like to understand how to stop a bot, any kind of help would be appreciated!

Thank you in advance!

**Reply by merschformann** [reply]
Unfortunately, the current bot agent implementation is not really prepared for simulating random bot errors. You are certainly looking at the right file though (_BotNormal.cs_).

However, the easiest lift / hack that comes to my mind may be stopping a robot randomly when it completes a partial hop of its path. That should make it a somewhat graceful stop from the simulations perspective and you don't need to worry too much about the reservation system, collision handling, etc.
You could probably do this somewhere around [here](https://github.com/merschformann/RAWSim-O/blob/0c217e36dcfef890a2bc79cbdb1f496484e2d46b/RAWSimO.Core/Bots/BotNormal.cs#L1107). Instead of allowing it to register it's next hop you could mark it as faulty and cancels it's move task. You could then add a fault task which could simulate some down-time before removing itself again.
This is only from the top of my head though. I would have to look at the code in more detail again to verify / find a better solution.

Of course, you can also spend more work on it and stop a robot's move task even while driving. Some more code may be required though. And don't forget to update the reservation system, since bots will crash otherwise. :relaxed: 

Hope this helps a little. And let me know what did not work with the PodStorage controller. Maybe I can help with that one.

P.S. hop == one straight drive action

**Reply by merschformann** [reply]
Addendum:
You mentioned that you are interested in the bot positions over time. They are typically written to a _locationspolled.heat_ file which is used for heatmap generation. I put together a small overview in a gif (see below). You can of course just use the bot positions written to the file directly. (_locationspolled.heat_ is also available via CLI)

![heatmap-quickstart](https://user-images.githubusercontent.com/11454283/124100801-0d9be180-da5f-11eb-8de8-4c1891bedfb0.gif)

Resulting heatmap from snapshot (snapshot is written to `<application dir>/snapshots`):

![1-4-4-32-428-2D-20210701112212](https://user-images.githubusercontent.com/11454283/124101485-b21e2380-da5f-11eb-9b87-9fba224e326b.png)

**Reply by LFK01** [reply]
Thank you for all the suggestions.

First of all, let me say that I tried again to build The TestPodStorageController and it worked fine. Previously, I had errors on some imports but I think it was because I was working in Folder View and not in Solution View (I didn't open the solution to modify it).

This morning I tried looking at the line you suggested and tried 2 modifications:

```
bool successfulRegistration = false;
```
without results; then I tried:
```
bool successfulRegistration = bot.setNextWaypoint(bot.CurrentWaypoint, currentTime);
```
and
```
bot.Path.Clear();
bool successfulRegistration = bot.setNextWaypoint(bot.Instance.Controller.PathManager.GetWaypointByNodeId(bot.Path.NextAction.Node), currentTime);
```
resulting in a crash of the program after the simulation start, if I remember correctly.

I know I need to dive deeper into the code and understand it better.

I don't understand very well what you mean by:

> You could then add a fault task which could simulate some down-time before removing itself again.

I saw the usage of the heatmap in your tutorial video and I thought some information could be extracted from those files but I haven't carefully looked into it yet.

Again, thank you for all the hints!

I'll keep trying to implement faults in the bots, if you have other suggestions I'm all ears.

**Reply by merschformann** [reply]
> I don't understand very well what you mean by:
>> You could then add a fault task which could simulate some down-time before removing itself again.

_Task_ was the wrong word, it should have been _state_. 😅Sorry for the confusion.

However, every bot manages a state queue. You can stop the robot in its `RAWSimO.Core.Bots.BotNormal.BotMove.Act(...)` method (that is where it is handling the move from A to B) either when it is standing still already (aforementioned idea) or put in some more logic for stopping gracefully (be aware that the reservation system needs to stay in sync). Then, you could insert a new state into the state queue which simulates the down-time and removes itself afterwards. If you do `bot.RequestReoptimization = true;` after down-time and the bot still has its destination, it should simply get a new path.
I think most of what you need to change for simulating bot errors can be done within `BotNormal.cs`. 

> I saw the usage of the heatmap in your tutorial video and I thought some information could be extracted from those files but I haven't carefully looked into it yet.

The `locationspolled.heat` file simply contains the polled positions of the bots over time. ☺️ 

I hope this helps a little. Sorry for the late reply, I didn't find much time earlier. 😬

**Reply by LFK01** [reply]
Hi Marius, thank you for your response.

> If you do `bot.RequestReoptimization = true;` after down-time and the bot still has its destination, it should simply get a new path.

helped me understand why the bot kept moving after telling them to stay in their `CurrentWaypoint`.

I managed to stop Bot 1 without crashing the simulator by substituting lines [274](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L274), [293](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L293), [322](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L322), [342](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L342), [459](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L459), [1034](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L1034), and [1102](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L1102) on `BotNormal.cs`:
```
RequestReoptimization = true;
```
and lines [1059](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L1059), [1107](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L1107):
```
successfulRegistration = bot.setNextWaypoint(bot.Instance.Controller.PathManager.GetWaypointByNodeId(bot.Path.NextAction.Node), currentTime);
```
with:
```
if(this.ID == 1) { RequestReoptimization = false; }
else { RequestReoptimization = true; }
```
and:
```
if (bot.ID == 1) {
	successfulRegistration = bot.setNextWaypoint(bot.CurrentWaypoint, currentTime);
	}
else {
	successfulRegistration = bot.setNextWaypoint(bot.Instance.Controller.PathManager.GetWaypointByNodeId(bot.Path.NextAction.Node), currentTime);
	}
```

For the moment, this should be enough to proceed with my thesis. I think it is not what you exactly meant in your replies but it is the best I could do right now. 
I tried to stop the bot after some timesteps by adding the condition `currentTime > 100` in the `if` clauses used above but I crashed the simulator again.
Now I'll focus on parsing the bot positions and processing the movement features.

If you have any other suggestions I'm always happy to hear them!

Thank you so much for the help you provided me!

**Reply by LFK01** [reply]
Hi Marius,

sorry to bother you again I have two things to ask.

First of all the easy one: I noticed that on the `locationspolled.heat` file there is a column for the task the bot is executing. In [`BotTaskType.cs`](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Control/BotTaskType.cs) there are 6 kind of tasks (None, ParkPod, RepositionsPod, Insert, Extract, Rest) while on the .heat file I extracted only 5 tasks `['N' 'E' 'I' 'S' 'P']`. I assumed N corresponds to None, E to Extract, I to Insert, P to ParkPod. What does the S stand for? And should there be another letter in the .heat file that I'm not seeing?

Secondly, I'm trying to implement a new fault since the first one is not quite realistic. I'd like to move a bot back and forth from two adjacents cells. At [1037](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L1037), I tried to retrieve two different Waypoints:
```
bot.first_position = bot.CurrentWaypoint;
bot.second_position = bot.Instance.Controller.PathManager.GetWaypointByNodeId(bot.Path.NextAction.Node);
```

I implemented this such that the bot gets a Path at first and then moves back and forth from the first two positions assigned. I assigned `bot.first_position` and `bot.second_position` at line [1059](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L1059) and at line [1107](https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L1107) with:
```
if (fault_bot)
{
	if (bot.first_position_done)
	{
		successfulRegistration = bot.setNextWaypoint(bot.second_position, currentTime);
		bot.first_position_done = false;
	}
	else
	{
		successfulRegistration = bot.setNextWaypoint(bot.first_position, currentTime);
		bot.first_position_done = true;
	}
}
else
{
        successfulRegistration = bot.setNextWaypoint(bot.Instance.Controller.PathManager.GetWaypointByNodeId(bot.Path.NextAction.Node), currentTime);
}
``` 

But I saw that:
 - CurrentWaypoint and NextAction.Node refers to the same position if called one after the other.
 - Even if I could manage to retrieve the next position to which `NextAction.Node` refers to, it would not be an adjacent cell but what I think would be the end of Path cell.

I need to specify that, for the purpose of my thesis, I don't need to concentrate on the completion of the orders but rather on the movements of the bots. 
I saw that, by blocking 20% of bots (pointed by red arrows in the picture), most of the working bots were resting by the conclusion of the simulation, could this be due to the allocation of tasks?
![vlcsnap-2021-07-15-22h44m33s637_LI](https://user-images.githubusercontent.com/37867917/125857800-10c93ec4-00d4-4961-a7eb-245029506ab9.jpg)
If this is a difficult problem to solve I could find a workaround by running shorter simulations or faulting fewer bots. 

Could you provide some hints on where I should look in order to understand the assignments of movements to the bots?

If I'm not clear or if I've said something wrong don't worry about pointing that out, there are a lot of things of which I'm not sure about.

Thank you in advance

**Reply by merschformann** [reply]
## Task abbreviations

The task types are defined here:

https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Statistics/DataPoint.cs#L3693-L3698

As you can see, the `S` stands for 'resting'. :sweat_smile: The letters collided. `R` is for repositioning a pod and won't show up, if not used. I used repositioning in this [paper](https://link.springer.com/chapter/10.1007/978-3-319-89920-6_51).
I had to save disk space when running my experiments back in the day, hence, the abbreviations.

## 

Regarding the method you are mentioning: keep in mind that it's called repeatedly and only updates the state. It's a little tricky to convince it of simulating down-time. Also, I don't think I understand the two-waypoint fail concept yet. :blush: 
However, if modifying the move handling gets too complicated, there may be another possibility by issuing additional stops or waiting when the move states are created. For this the following code should be interesting:

https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L380-L393

If you need different behavior depending on the outer task, the code below could be interesting too (the one calling above code):

https://github.com/merschformann/RAWSim-O/blob/df4874c9f0591954c312ad351b27cda8a42cecd6/RAWSimO.Core/Bots/BotNormal.cs#L267-L368

All bots resting is weird and shouldn't happen, if you still have inventory, orders and a default bot manager, etc. Although, it's quite hard though to see what's going on without any code, etc. :sweat_smile: 
I fully understand that figuring out the dynamics of the bot movement simulation is not easy. I can recommend carefully stepping through it with a debugger to see how it behaves and where you can hijack it for your purpose. :relaxed:

**Comment by Mitchel703** [comment]
Dear Marius, 
I would like to be able to more precisely adjust the layout of the pods, i.e. having different sized locations in a single pod (volume wise), instead of only being able to adjust the total 'weight'. Do you know if this is possible and if so, do you have any tips on how to code this properly? Thank you in advance!

**Reply by merschformann** [reply]
Hi @Mitchel703!

The capacity of a pod is only one-dimensional as you have noticed. It is possible to have differently capacitated pods though.

However, it seems like you are asking for multiple heterogeneously sized storage compartments on one pod. :thinking: 

When doing this properly the following steps should roughly outline what needs to be changed:

1. Change the [Pod](https://github.com/merschformann/RAWSim-O/blob/main/RAWSimO.Core/Elements/Pod.cs) model to account for the storage compartments. Most importantly, the pod capacity & capacity in use management needs to change. I would personally start there and work my way backwards through the rest of the code.
2. If you don't want to rely on a fixed rule for selecting the storage compartment in a pod, I would suggest you think about making that decision logic configurable. I.e., there are 3 compartments in a pod that can accept an item bundle, which one should we use? (the one with least residual space?) This is more advanced though and may be skipped in a first attempt. If you aim to do this, look into the other decision managers. It should be somewhat straightforward to add a similar configuration + managers structure that you can vary in experiments to compare their performance.
3. Check whether the logic in the `BundleManager`s needs to be adapted. They [choose which pod to use](https://github.com/merschformann/RAWSim-O/blob/main/RAWSimO.Core/Control/BundleManager.cs#LL50C61-L50C61) per item bundle that needs to be stored. It seems like they use the [FitsForReservation](https://github.com/merschformann/RAWSim-O/blob/main/RAWSimO.Core/Elements/Pod.cs#L271-L276) method to determine whether the bundle fits the pod, so, they may already work with the changes done in 1. It could be though that some kind of integration of the two decision problems needs to be done to achieve good results.

Other questions:
- Does a compartment need to be SKU-clean? Or can two different products share the same one? I assumed above that they can be mixed. Should mixing be avoided if possible? This needs to be reflected in the compartment decision logic mentioned above in point 2.

I hope this helps a bit. Feel free to come back with follow-up questions. :relaxed: 

I hope RAWSim-O can be helpful to your project! :hugs:

**Reply by Mitchel703** [reply]
Dear Marius,

First of all thank you for your response. To answer your question, for the experiments I would like to perform the compartments need to be SKU clean indeed. The point is that I would like to investigate the difference in total used storage capacity when having heterogenuous compartments instead homogenuous.  Furthermore I should mention that my coding experience and especially C# experience is very very low. I believe I understand where to look and what needs to be changed (atleast for step 1 and 3) however I am unsure how to change it. So it would be very helpful if you could assist me with a small example of how I should adjust the code or maybe a bit more guidance codewise. I hope you can find the time to help me a bit more, anyway thank you very much in advance for your response!

Kind regards,

Mitchel

**Reply by merschformann** [reply]
Hey Mitchel!

Sorry for the late reply, I currently have a lot going on. :sweat_smile: 
It would be difficult for me to support with code right now. Feel free to ping me with questions though. It may take me a bit, but I normally reply sooner or later. :relaxed: 

For a quicker back & forth you could ping me on an old discord server I should be idling frequently. Find it [here](https://discord.gg/YGNWJAC32J).

Kind regards,
Marius

**Comment by lsq7773** [comment]
Hi Marius！
Thank you for your work. It's quite a good job which helped me a lot.
However, due to my poor knowledge base in coding and algorithm, I feel it hard for me to understand the specific function of the buttons in _Controller configuration_. 
So, could you please release some specific introductions like manual? I will be grateful.

**Reply by merschformann** [reply]
Hi @lsq7773 !

Could you elaborate a bit where you exactly get stuck? Is it about specific settings in _Controller Configuration_ or about getting something to run in general?

There is a cringy video tutorial about the basics around getting started [here](https://youtu.be/l3qwDnbEPhI). And there is also some more information in the [wiki](https://github.com/merschformann/RAWSim-O/wiki). Furthermore, the simulation concept is described in [this paper](https://arxiv.org/pdf/1710.04726.pdf) and most basic controllers in [this paper](https://www.sciencedirect.com/science/article/pii/S2214716019300946). 

It's a bit scattered to be frank, but maybe it helps you already. Otherwise, let me know what you are exactly looking for and I will try to find some time and provide a better introduction.

Kind regards,
Marius

**Reply by lsq7773** [reply]
Hi Marius！
Thank you so much for the paper you provided to me. I’ve been reading them these days. It helped me a lot. 
Now I met some new problems.  
1.  I’m not clear how the module _PodSelectionConfig_ under the _BalancedTaskAllocationConfig_ works. In other words, I am not clear about the function of _TaskAllocationConfig_ and didn't find specific explanation from the resource above. Now I want to make the loaded bots travel to the least crowded pickstation. Can I adjust the parameters of this module to realize the function?
2.  With the need to do controlled trial, I also want to apply an order task which can be set and fixed, whose data can eventually be extracted. 
3. Finally, I wonder where to embed my code for robot scheduling rules.
As a beginner, I know I'm asking a lot of low-level questions. I hope I didn't cause you too much trouble. 

Thank you in advance.
